src/sagemaker/model.py

```diff
@@ -242,10 +394,90 @@ def prepare_container_def(
         Returns:
             dict: A container definition object usable with the CreateModel API.
         """
+        deploy_key_prefix = fw_utils.model_code_key_prefix(
+            self.key_prefix, self.name, self.image_uri
+        )
+        deploy_env = copy.deepcopy(self.env)
+        if self.source_dir or self.dependencies or self.entry_point or self.git_config:
+            if self.key_prefix or self.git_config:
+                self._upload_code(deploy_key_prefix, repack=False)
+            elif self.source_dir and self.entry_point:
+                self._upload_code(deploy_key_prefix, repack=True)
+            else:
+                self._upload_code(deploy_key_prefix, repack=False)
+            deploy_env.update(self._script_mode_env_vars())
         return sagemaker.container_def(
-            self.image_uri, self.model_data, self.env, image_config=self.image_config
+            self.image_uri, self.model_data, deploy_env, image_config=self.image_config
         )
 
+    def _upload_code(self, key_prefix: str, repack: bool = False) -> None:
+        """Uploads code to S3 to be used with script mode with SageMaker inference.
+
+        Args:
+            key_prefix (str): The S3 key associated with the ``code_location`` parameter of the
+                ``Model`` class.
+            repack (bool): Optional. Set to ``True`` to indicate that the source code and model
+                artifact should be repackaged into a new S3 object. (default: False).
+        """
+        local_code = utils.get_config_value("local.local_code", self.sagemaker_session.config)
+        if (self.sagemaker_session.local_mode and local_code) or self.entry_point is None:
+            self.uploaded_code = None
+        elif not repack:
+            bucket = self.bucket or self.sagemaker_session.default_bucket()
+            self.uploaded_code = fw_utils.tar_and_upload_dir(
+                session=self.sagemaker_session.boto_session,
+                bucket=bucket,
+                s3_key_prefix=key_prefix,
+                script=self.entry_point,
+                directory=self.source_dir,
+                dependencies=self.dependencies,
+            )
+
+        if repack and self.model_data is not None and self.entry_point is not None:
+            if isinstance(self.model_data, sagemaker.workflow.properties.Properties):
+                # model is not yet there, defer repacking to later during pipeline execution
+                return
```

Comment:
This feels similar to elsewhere in the code, can we in future pull this into a common utility?

Reply:
We have this logic only once in the code, though there's similar packaging code used for the `Estimator` classes.